stability: crash in timestamp cache

Build sha: 6188995f5f0868ee57378a0e7e507dea80e3cd38

Full log (through panic parse): 
[node4.crash.parsed.txt](https://github.com/cockroachdb/cockroach/files/159885/node4.crash.parsed.txt)

Relevant bit:
```
I160304 23:40:33.117798 gossip/gossip.go:941  starting client to ec2-54-208-199-194.compute-1.amazonaws.com:26257
panic: start key greater than or equal to end key ^Dtsd!cr.node.exec.error-count^@^A<89>ø^F-^X >= ^Dtsd!cr.node.exec.error-count^@^A<89>ø^F-^X


To see all goroutines, visit https://github.com/maruel/panicparse#GOTRACEBACK

1: running [Created by rpc.syncAdapter.exec @ server.go:51]
    cache   cache.go:434           (*IntervalCache).MakeKey(0xc82058d500, #4, 0x24, 0x30, #5, 0x24, 0x30, 0, 0, 0, ...)
    storage timestamp_cache.go:128 (*TimestampCache).Add.func1(#4, 0x24, 0x30, #5, 0x24, 0x30)
    storage timestamp_cache.go:245 (*TimestampCache).Add(0xc82036d6b0, 0xc82cf64fc0, 0x24, 0x30, #5, 0x24, 0x30, #9, 0, 0, ...)
    storage replica.go:769         (*Replica).endCmds(#2, 0xc852800180, 0x1, 0x1, 0xc8211a8900, 0)
    storage replica.go:749         (*Replica).beginCmds.func1(0)
    storage replica.go:853         (*Replica).addReadOnlyCmd.func1(0xc84a3c5530, 0xc8422beec8)
    storage replica.go:869         (*Replica).addReadOnlyCmd(#2, #8, #6, #9, 0, #1, 0x9, 0x1, 0, 0, ...)
    storage replica.go:655         (*Replica).Send(#2, #8, #6, #9, 0, #1, 0x9, 0x1, 0, 0, ...)
    storage store.go:1450          (*Store).Send(0xc820296000, #8, #6, #9, 0, #1, 0x9, 0x1, 0, 0, ...)
    storage stores.go:186          (*Stores).Send(0xc820174f40, #8, 0xc8319b8a50, 0, 0, #1, 0x9, 0x1, 0, 0, ...)
    server  node.go:607            (*Node).executeCmd.func2()
    stop    stopper.go:125         (*Stopper).RunTask(0xc8202b3ab0, 0xc8422bfe98, 0)
    server  node.go:619            (*Node).executeCmd(0xc8204ce180, #7, #3, 0, 0, 0, 0)
    server  node.go:326            executeCmd)-fm(#7, #3, 0, 0, 0, 0)
    rpc     server.go:50           syncAdapter.exec.func1(0xc8368f8500, 0xc8203413b0, #7, #3)
```

--- a/storage/timestamp_cache.go
+++ b/storage/timestamp_cache.go
@@ -158,8 +158,8 @@ func (tc *TimestampCache) Add(start, end roachpb.Key, timestamp roachpb.Timestam
 				case sCmp <= 0 && eCmp >= 0:
 					// New contains or is equal to old; delete old.
 					//
-					// New: ------------
-					// Old:   --------
+					// New: ------------      ------------      ------------
+					// Old:   --------    or    ----------  or  ----------
 					//
 					// Old:
 					cache.DelEntry(o.Entry)
@@ -179,18 +179,18 @@ func (tc *TimestampCache) Add(start, end roachpb.Key, timestamp roachpb.Timestam
 				case eCmp >= 0:
 					// Left partial overlap; truncate old end.
 					//
-					// New:     --------
-					// Old: --------
+					// New:     --------          --------
+					// Old: --------      or  ------------
 					//
-					// Old: ----
+					// Old: ----              ----
 					o.Key.End = r.Start
 				case sCmp <= 0:
 					// Right partial overlap; truncate old start.
 					//
-					// New: --------
-					// Old:     --------
+					// New: --------          --------
+					// Old:     --------  or  ------------
 					//
-					// Old:         ----
+					// Old:         ----              ----
 					o.Key.Start = r.End
 				default:
 					panic(fmt.Sprintf("no overlap between %v and %v", o.Key.Range, r))
@@ -203,8 +203,8 @@ func (tc *TimestampCache) Add(start, end roachpb.Key, timestamp roachpb.Timestam
 				case sCmp >= 0 && eCmp <= 0:
 					// Old contains or is equal to new; no need to add.
 					//
-					// Old: ------------
-					// New:    ------
+					// Old: -----------      -----------      -----------      -----------
+					// New:    -----     or  -----------  or  --------     or     --------
 					//
 					// New:
 					return
@@ -221,21 +221,21 @@ func (tc *TimestampCache) Add(start, end roachpb.Key, timestamp roachpb.Timestam
 					addRange(lr)
 
 					r.Start = o.Key.End
-				case eCmp >= 0:
+				case eCmp > 0:
 					// Left partial overlap; truncate new start.
 					//
-					// Old: --------
-					// New:     --------
+					// Old: --------          --------
+					// New:     --------  or  ------------
 					//
-					// New:         ----
+					// New:         ----              ----
 					r.Start = o.Key.End
-				case sCmp <= 0:
+				case sCmp < 0:
 					// Right partial overlap; truncate new end.
 					//
-					// Old:     --------
-					// New: --------
+					// Old:     --------          --------
+					// New: --------      or  ------------
 					//
-					// New: ----
+					// New: ----              ----
 					r.End = o.Key.Start
 				default:
 					panic(fmt.Sprintf("no overlap between %v and %v", o.Key.Range, r))

--- a/storage/timestamp_cache_test.go
+++ b/storage/timestamp_cache_test.go
@@ -360,6 +360,90 @@ var layeredIntervalTestCase2 = layeredIntervalTestCase{
 	},
 }
 
+// layeredIntervalTestCase3 tests a right partial overlap with a shared end
+// for adding intervals to the interval cache when tested in order, and
+// tests a left partial overlap with a shared end when tested in reverse.
+var layeredIntervalTestCase3 = layeredIntervalTestCase{
+	actions: []func(tc *TimestampCache, ts roachpb.Timestamp){
+		func(tc *TimestampCache, ts roachpb.Timestamp) {
+			// No overlap forwards.
+			// Right partial overlap backwards.
+			tc.Add(roachpb.Key("a"), roachpb.Key("c"), ts, nil, true)
+		},
+		func(tc *TimestampCache, ts roachpb.Timestamp) {
+			// Left partial overlap forwards.
+			// No overlap backwards.
+			tc.Add(roachpb.Key("b"), roachpb.Key("c"), ts, nil, true)
+		},
+	},
+	validator: func(t *testing.T, tc *TimestampCache, tss []roachpb.Timestamp) {
+		acTS := tss[0]
+		bcTS := tss[1]
+
+		// Try different sub ranges.
+		if rTS := tc.GetMaxRead(roachpb.Key("a"), nil, nil); !rTS.Equal(acTS) {
+			t.Errorf("expected \"a\" to have acTS %v timestamp, found %v", acTS, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("b"), nil, nil); !rTS.Equal(bcTS) {
+			t.Errorf("expected \"b\" to have bcTS %v timestamp, found %v", bcTS, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("c"), nil, nil); !rTS.Equal(tc.lowWater) {
+			t.Errorf("expected \"c\" to have low water %v timestamp, found %v", tc.lowWater, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("a"), roachpb.Key("c"), nil); !rTS.Equal(bcTS) {
+			t.Errorf("expected \"a\"-\"c\" to have bcTS %v timestamp, found %v", bcTS, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("a"), roachpb.Key("b"), nil); !rTS.Equal(acTS) {
+			t.Errorf("expected \"a\"-\"b\" to have acTS %v timestamp, found %v", acTS, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("b"), roachpb.Key("c"), nil); !rTS.Equal(bcTS) {
+			t.Errorf("expected \"b\"-\"c\" to have bcTS %v timestamp, found %v", bcTS, rTS)
+		}
+	},
+}
+
+// layeredIntervalTestCase4 tests a left partial overlap with a shared start
+// for adding intervals to the interval cache when tested in order, and
+// tests a right partial overlap with a shared start when tested in reverse.
+var layeredIntervalTestCase4 = layeredIntervalTestCase{
+	actions: []func(tc *TimestampCache, ts roachpb.Timestamp){
+		func(tc *TimestampCache, ts roachpb.Timestamp) {
+			// No overlap forwards.
+			// Left partial overlap backwards.
+			tc.Add(roachpb.Key("a"), roachpb.Key("c"), ts, nil, true)
+		},
+		func(tc *TimestampCache, ts roachpb.Timestamp) {
+			// Right partial overlap forwards.
+			// No overlap backwards.
+			tc.Add(roachpb.Key("a"), roachpb.Key("b"), ts, nil, true)
+		},
+	},
+	validator: func(t *testing.T, tc *TimestampCache, tss []roachpb.Timestamp) {
+		acTS := tss[0]
+		abTS := tss[1]
+
+		// Try different sub ranges.
+		if rTS := tc.GetMaxRead(roachpb.Key("a"), nil, nil); !rTS.Equal(abTS) {
+			t.Errorf("expected \"a\" to have abTS %v timestamp, found %v", abTS, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("b"), nil, nil); !rTS.Equal(acTS) {
+			t.Errorf("expected \"b\" to have acTS %v timestamp, found %v", acTS, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("c"), nil, nil); !rTS.Equal(tc.lowWater) {
+			t.Errorf("expected \"c\" to have low water %v timestamp, found %v", tc.lowWater, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("a"), roachpb.Key("c"), nil); !rTS.Equal(abTS) {
+			t.Errorf("expected \"a\"-\"c\" to have abTS %v timestamp, found %v", abTS, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("a"), roachpb.Key("b"), nil); !rTS.Equal(abTS) {
+			t.Errorf("expected \"a\"-\"b\" to have abTS %v timestamp, found %v", abTS, rTS)
+		}
+		if rTS := tc.GetMaxRead(roachpb.Key("b"), roachpb.Key("c"), nil); !rTS.Equal(acTS) {
+			t.Errorf("expected \"b\"-\"c\" to have acTS %v timestamp, found %v", acTS, rTS)
+		}
+	},
+}
+
 // TestTimestampCacheLayeredIntervals verifies the maximum timestamp
 // is chosen if previous entries have ranges which are layered over
 // each other.
@@ -378,6 +462,8 @@ func TestTimestampCacheLayeredIntervals(t *testing.T) {
 	for _, testCase := range []layeredIntervalTestCase{
 		layeredIntervalTestCase1,
 		layeredIntervalTestCase2,
+		layeredIntervalTestCase3,
+		layeredIntervalTestCase4,
 	} {
 		// Perform actions in order and validate.
 		tc.Clear(clock)

